tao_compiler/mlir/disc/transforms/disc_specialize_fusion_with_speculation.cc

```diff
@@ -262,28 +258,65 @@ struct DiscSpecializeFusionWithSpeculationPass
     Operation* reduce_op = GetCandidateRowReduceOp(fusion_op);
     assert(reduce_op != nullptr);
 
-    // Row reduction schedule selection policy: if col size > 512, use schedule
-    // 1; also use schedule 2. The 512 is an empirical value.
     OpBuilder b(fusion_op);
     Location loc = fusion_op.getLoc();
     FusionOp cloned = dyn_cast<FusionOp>(b.clone(*fusion_op.getOperation()));
 
     Value operand = reduce_op->getOperand(0);
+    // TODO(disc): Use 256 as default block size; turn this number for
+    // different shapes
+    int block_size = kThreadsRowReduction;
     Value col_size = b.create<memref::DimOp>(loc, operand, 1);
-    Value ref_size =
-        b.create<arith::ConstantIndexOp>(loc, kRowReductionScheduleTurningSize);
-    Value pred = b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::sgt,
-                                         col_size, ref_size);
+    Value pred;
+
+    // TODO: this feature will be experimental in the first release, and will be
+    // set as default in the near future after evaluated on benchmarks. And
+    // later, the two speculations will be merge into one according to the
+    // adaptive thread mapping approach in AStitch. Then there will be more
+    // flexible per-reduce-thread-number setting.
+    bool experimental_tlp_enhance = false;
+    tensorflow::ReadBoolFromEnvVar("DISC_EXPERIMENTAL_SPECULATION_TLP_ENHANCE",
+                                   false, &experimental_tlp_enhance);
+    auto thread_number_info =
+        ArchToGPUThreadNumber.find({cc_major_, cc_minor_});
+    if (experimental_tlp_enhance &&
+        (thread_number_info != ArchToGPUThreadNumber.end())) {
+      // Experimental schedule selection policy is as following:
+      //   1. use schedule 1 if
```

Comment:
Is 'one block per row' better when both row size and column size are small?

Reply:
Neither of existing 2 schedules can deal with the case you mentioned. This is why I leave a TODO in the comments about re-design the schedule according to `adaptive thread mapping` in AStitch.